src/ui/public/visualize/loader.js

```diff
@@ -1,75 +1,170 @@
-import $ from 'jquery';
-import uiRoutes from 'ui/routes';
+/**
+ * IMPORTANT: If you make changes to this API, please make sure to check that
+ * the docs (docs/development/visualize/development-create-visualization.asciidoc)
+ * are up to date.
+ */
+import angular from 'angular';
+import chrome from 'ui/chrome';
 import 'ui/visualize';
+import visTemplate from './loader_template.html';
+
+
+/**
+ * A handler to the embedded visualization. It offers several methods to interact
+ * with the visualization.
+ */
+class EmbeddedVisualizeHandler {
+  constructor(element, scope) {
+    this._element = element;
+    this._scope = scope;
+    this._renderComplete = new Promise(resolve => {
+      this._element.on('renderComplete', resolve);
```

Comment:
This relies on the assumption visualizations only fire renderComplete once. Which they don't. (e.g. maps are interactive, and modify the UI-state, rerending).

I think we can make this assumption in this context (people need to embed and are just interested in the first "success" status), but then we should immediately remove the listener after resolving.

Reply:
Hmm good point. I am not sure if people will just need to listen for the first render. So I would actually check with some people actually already using this API, to see what their use cases would be. I would either use the promise, rename the method and only listen once, or alternatively implement a proper callback mechanism to register and deregister for `renderComplete` events. But I would rather have people to use methods of this API, than needing to listen for the hardcoded string themselves, which will make it harder for us later to change implementations.

Reply:
Maybe we could also for convenience do both, offer the `whenFirstRenderComplete()` promise and the event mechanism.

Reply:
@kobelb - care to pop in here and just make sure the renderComplete stuff should work?  I confirmed the tests for print layout pass, and that reports with TSVB and timelion are generated as expected.  I *think* that would trigger an issue if render complete wasn't being counted accurately, but I'm not 100% sure.